test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,769 @@
+            persistElement(moduleElement, moduleElement, ct.getTypes(), releaseVersion);
+            if (ed.getTargetModules() == null) {
+                analyzePackageRecord(ed.getPackage(), releaseVersion, ct);
+            }
+        }
+    }
+
+    private void analyzePackageRecord(PackageElement pe, String releaseVersion, JavacTask ct) {
+        persistElement(pe, pe, ct.getTypes(), releaseVersion);
+        List<TypeElement> typeElements = ElementFilter.typesIn(pe.getEnclosedElements());
+        for (TypeElement te : typeElements) {
+            analyzeClassRecord(te, releaseVersion, ct.getTypes(), ct.getElements());
+        }
+    }
+
+    private void analyzeClassRecord(TypeElement te, String version, Types types, Elements elements) {
+        Set<Modifier> classModifiers = te.getModifiers();
+        if (!(classModifiers.contains(Modifier.PUBLIC) || classModifiers.contains(Modifier.PROTECTED))) {
+            return;
+        }
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(element -> element.getModifiers().contains(Modifier.PUBLIC) || element.getModifiers().contains(Modifier.PROTECTED))
+                .filter(element -> element.getKind().isField()
+                        || element.getKind() == ElementKind.METHOD
+                        || element.getKind() == ElementKind.CONSTRUCTOR)
+                .forEach(element -> persistElement(te, element, types, version));
+        te.getEnclosedElements().stream()
+                .filter(element -> element.getKind().isDeclaredType())
+                .map(TypeElement.class::cast)
+                .forEach(nestedClass -> analyzeClassRecord(nestedClass, version, types, elements));
+    }
+
+    public void persistElement(Element explicitOwner, Element element, Types types, String version) {
+        String uniqueId = getElementName(explicitOwner, element, types);
+        IntroducedIn introduced = classDictionary.computeIfAbsent(uniqueId, i -> new IntroducedIn());
+        if (isPreview(element, uniqueId, version)) {
+            if (introduced.introducedPreview == null) {
+                introduced.introducedPreview = version;
+            }
+        } else {
+            if (introduced.introducedStable == null) {
+                introduced.introducedStable = version;
+            }
+        }
+    }
+
+    private boolean isPreview(Element el, String uniqueId, String currentVersion) {
+        while (el != null) {
+            Symbol s = (Symbol) el;
+            if ((s.flags() & Flags.PREVIEW_API) != 0) {
+                return true;
+            }
+            el = el.getEnclosingElement();
+        }
+
+        return LEGACY_PREVIEW_METHODS.containsKey(currentVersion)
+                && LEGACY_PREVIEW_METHODS.get(currentVersion).contains(uniqueId);
+    }
+
+    private void testThisModule(String moduleName) throws Exception {
```

Comment:
The word `this` is a bit distracting/unnecessary here.  It's enough to say `testModule(String moduleName)` although `checkModule(String moduleName)` would be even better

Reply:
Thanks, renamed it to `CheckModule`